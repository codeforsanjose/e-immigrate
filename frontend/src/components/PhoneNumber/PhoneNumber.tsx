import React from 'react';
import './PhoneNumber.css';
import { ReactSetter } from '../../types/common';
import { CommonComponentProps } from '../../types/CommonComponentProps';
import { useContentContext } from '../../contexts/ContentContext';
import { RequiredErrorDiv } from '../RequiredErrorPresenter';
import { useQuestionnaireResponseContext } from '../../contexts/QuestionnaireResponseContext';

type PhoneNumberProps = CommonComponentProps & {
    setErrors: ReactSetter<Record<string, unknown>>;
};
export function PhoneNumber(props: PhoneNumberProps) {
    const {
        q, 
        setErrors, 
    } = props;
    const { content } = useContentContext();
    const { 
        collectAnswer,
        bindField, 
    } = useQuestionnaireResponseContext();
    const onChange = React.useCallback((e: React.ChangeEvent<HTMLInputElement>) => {
        const isValid = e.target.checkValidity();
        setErrors(prev => {
            if (prev[q.slug] === isValid) return prev;
            return { 
                ...prev, 
                [q.slug]: isValid,
            };
        });
        collectAnswer(q.slug, e.target.value);
    }, [collectAnswer, q.slug, setErrors]);
    return (
        <>
            <input
                type="tel"
                pattern="^(\+\d{1,2}\s)?\(?\d{3}\)?[\s.\-]?\d{3}[\s.\-]?\d{4}$"
                id={q.slug}
                name={q.slug}
                required={q.required}
                className="TextInput"
                {...bindField(q.slug)}
                onChange={onChange} />
            <RequiredErrorDiv message={content.errorMessagePhone} />
        </>
    );
}
