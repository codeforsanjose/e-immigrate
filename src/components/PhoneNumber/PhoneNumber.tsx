import React from 'react';
import './PhoneNumber.css';
import { BindFieldFunction, CollectAnswerFunction, QData, ReactSetter } from '../../types/common';




type PhoneNumberProps = {
    attributes: {
        q: QData;
        bindField: BindFieldFunction;
        collectAnswer: CollectAnswerFunction;
        setErrors: ReactSetter<Record<string, unknown>>;
        content: {
            errorMessagePhone: string;
        };
    }
};
export function PhoneNumber(props: PhoneNumberProps) {
    const {
        attributes: { 
            q, 
            bindField, 
            collectAnswer, 
            setErrors, 
            content,
        },
    } = props;
    return (
        <>
            <input
                type="tel"
                pattern="^(\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4}$"
                id={q.slug}
                name={q.slug}
                required={q.required}
                className="TextInput"
                {...bindField(q.slug)}
                onChange={(e) => {
                    if (e.target.checkValidity()) {
                        setErrors((prev) => ({ ...prev, [q.slug]: true }));
                    } else {
                        setErrors((prev) => ({ ...prev, [q.slug]: false }));
                    }
                    collectAnswer(q.slug, e.target.value);
                } } />
            <div className="RequiredError">*{content.errorMessagePhone}</div>
        </>
    );
}
