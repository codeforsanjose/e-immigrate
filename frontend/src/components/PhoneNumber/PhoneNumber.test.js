import React from 'react';
import PhoneNumber from './PhoneNumber';

import renderer from 'react-test-renderer';

it('PhoneNumber renders correctly', () => {
    const tree = renderer
        .create(
            <PhoneNumber
                q={{}}
                answers={[]}
                values={[]}
                bindField={() => {}}
                showFollowUp={{}}
                setShowFollowUp={{}}
                collectAnswer={{}}
                content={{}}
            />
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});
