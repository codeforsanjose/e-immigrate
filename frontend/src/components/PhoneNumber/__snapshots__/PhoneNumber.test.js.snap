// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhoneNumber renders correctly 1`] = `
Array [
  <input
    className="TextInput"
    onChange={[Function]}
    pattern="^(\\\\+\\\\d{1,2}\\\\s)?\\\\(?\\\\d{3}\\\\)?[\\\\s.-]?\\\\d{3}[\\\\s.-]?\\\\d{4}$"
    type="tel"
  />,
  <div
    className="RequiredError"
  >
    *
  </div>,
]
`;
